lib/dry/system/importer.rb

```diff
@@ -38,15 +53,74 @@ def [](name)
       def key?(name)
         registry.key?(name)
       end
+      alias_method :namespace?, :key?
 
       # @api private
-      def call(ns, other)
-        container.merge(other, namespace: ns)
+      def finalize!
+        registry.each_key { import(_1) }
+        self
       end
 
       # @api private
-      def register(other)
-        registry.update(other)
+      def import(namespace, keys: Undefined)
+        item = self[namespace]
+        keys = Undefined.default(keys, item.import_keys)
+
+        if keys
+          import_keys(item.container, namespace, keys_to_import(keys, item))
+        else
+          import_all(item.container, namespace)
+        end
+
+        self
+      end
+
+      private
+
+      def keys_to_import(keys, item)
+        keys
+          .then { (arr = item.import_keys) ? _1 & arr : _1 }
+          .then { (arr = item.container.exports) ? _1 & arr : _1 }
+      end
+
+      def import_keys(other, namespace, keys)
+        container.merge(build_merge_container(other, keys), namespace: namespace)
+      end
+
+      def import_all(other, namespace)
+        merge_container =
+          if other.exports
+            build_merge_container(other, other.exports)
+          else
+            build_merge_container(other.finalize!, other.keys)
+          end
+
+        container.merge(merge_container, namespace: namespace)
+      end
+
+      def build_merge_container(other, keys)
+        keys.each_with_object(Dry::Container.new) { |key, ic|
+          next unless other.key?(key)
+
+          # Access the other container's items directly so that we can preserve all their
+          # options when we merge them with the target container (e.g. if a component in
+          # the provider container was registered with a block, we want block registration
+          # behavior to be exhibited when later resolving that component from the target
+          # container).
+          item = other._container[key]
```

Comment:
Could you add a TODO here that it probably indicates a missing API in dry-container?

Reply:
Will do. We actually do it in a couple of places in dry-system now, so it does feel like a part of dry-container worth expanding upon (speaking of which, we probably need to bring _that_ to 1.0 soon too).